src/main/java/org/folio/edge/sip2/parser/Parser.java

```diff
@@ -246,9 +248,16 @@ private ErrorDetection validateChecksum(String message, Command command) {
   }
 
   private Command parseCommandIdentifier(String message) {
-    String commandString = message.substring(0, 2);
-    Command command = Command.find(commandString);
+    final Command command;
+
+    if (message != null && message.length() >= 2) {
```

Comment:
Should this logic be in Command.find() itself? Or have we encountered situations where message == null or messageLength<2 to make it blow up on the spot so we have to check it first before issuing Command.find()?

Reply:
We could pass in the whole messsage to Command.find() to and this logic of checking for null and length >=2 can be packaged in there.

Reply:
Yes, the problem here was that `message.substring(0, 2)` was throwing exceptions, either NPE or index exceptions when bogus data was received.

My problem with passing the whole message to `Command.find()` is that extracting the command from the message string is the job the parser. In my mind, the enumeration should be kept as simple as possible, it should be able to map a command string to a `Command` enum and that's it.

Does this look better?

```suggestion
    command = Command.find(Optional.ofNullable(message)
        .filter(m -> m.length() >= 2)
        .map(m -> m.substring(0, 2))
        .orElse(null));
```

It is pretty much the equivalent, but we are allowing `Command.find()` to do the lookup and no logic is in this class for which command to assign when the message is wonky.

Reply:
OK, I see what you mean about the parser's responsibility. Either way is fine, I don't mind that "if" statement, to be honest, as I find it more readable, but it shouldn't make a difference to someone who is fluent in Java 8.

Reply:
How about this?:
```
    command = Command.find(message != null && message.length() >= 2 ? message.substring(0, 2) : null);
```

Reply:
To be honest, I like the `Optional` version better. 😄

Reply:
Actually, what I should do is move the check from the `parseCommandIdentifier` to `parseMessage` and do this check before calling `parseCommandIdentifier`.

Reply:
I like this idea. I'm curious how it will play out because the outcome of the check returns a Command.

Reply:
I think if we check before and find the message is null (I'm not sure it is possible) or message < 2 (which can happen) we just return a `Command.UNKNOWN` Message and short circuit the parsing. In this case, there is no way to parse a command or validate the message or anything else the parser does.